### PR TITLE
dump: Add an interaction option to the mermaid result.

### PR DESCRIPTION
**Pull request by wendyisdream**
This adds an interaction option to mermaid results. It allows users to
 check the call graph step by step with one uftrace result.
It supports the following graph following the user's input.
  a) Show the graph until specific call depth level.
     When the function has sub call flow, clicking the function node
      makes it attach the remaining call flow.
     It allows to toggle show/hide sub-graph.
  b) Show the partial graph starting from a specific function.
     This partial graph also support the a) depth option

For this feature, the following is added.
  1. Add the call depth level to mermaid node id as "depth_id" format.
  2. Add html and javascript to make the option. This code is included
      in the header file as raw string.
     This header file is converted incorrectly in pre-commit
      clang-format hook. So it is excluded in pre-commit.

Signed-off-by: Jia Ha <wendyisdream1004@gmail.com>

### CODE REVIEW

**Comment by wendyisdream** [comment]
Below is this option's operation example.
example_mermaid_option

**Comment by namhyung** [comment]
Thanks for this work.  Looks very useful!  I'll take a look later.

File: cmds/dump.c [hunk 1]
```diff
@@ -7,6 +7,7 @@
 
 #include "libtraceevent/event-parse.h"
 #include "libtraceevent/kbuffer.h"
+#include "mermaid.h"
```
Comment (namhyung) [inline cmds/dump.c:10]: I'd like to move the header in the utils directory.

File: tests/runtest.py [hunk 2]
```diff
@@ -449,17 +449,17 @@ def mermaid_sort(self, output, ignore_children=False):
...
+            if ln.find('<div class=\"mermaid\"') >= 0:
                 start_mermaid = True
                 continue
             if start_mermaid == False:
                 continue
             if ln.find('</div>') >= 0:
                 break
 
-            m = re.match( r'\s+(?P<start_id>\d+)\[\"(?P<start_name>\S+)\"\]\s+-->\|(?P<call_num>\d+)\|\s+(?P<end_id>\d+)\[\"(?P<end_name>\S+)\"\];', ln)
+            m = re.match( r'\s+(?P<start_depth>\d+)_(?P<start_id>\d+)\[\"(?P<start_name>\S+)\"\]\s+-->\|(?P<call_num>\d+)\|\s+(?P<end_depth>\d+)_(?P<end_id>\d+)\[\"(?P<end_name>\S+)\"\];', ln)
```
Comment (namhyung) [inline tests/runtest.py:460]: Could you please this line shorter?  You can break it in multiple lines if possible.

File: cmds/mermaid.h [hunk 3]
```diff
@@ -0,0 +1,256 @@
+const char *INTERACTIVE_OPTION_HTML = R""(
```
Comment (honggyukim) [inline cmds/mermaid.h:1]: I think this header is not needed. Instead, the contents of this variable can be in a separate file such as `mermaid.html` and it can be written as follows and `cmds/dump.c` can directly use it.

In `cmds/dump.c`,
```c
const char *interactive_option_html =
#include "utils/mermaid.html"
;
```
In `utils/mermaid.html`,
```html
R""(
<style>
#options {
  border-collapse: collapse;
  width: 100%;
}
    ... (put all the remaining contents html related) ...
</table>
<br>
</details>
)""
```

File: cmds/mermaid.h [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+          <option value="7">7</option>
+        </select>
+    </td>
+  </tr>
+</table>
+<br>
+</details>
+)"";
+
+const char *INTERACTIVE_OPTION_JS = R""(
```
Comment (honggyukim) [inline cmds/mermaid.h:61]: This part can also be written similarly as follows.

In `cmds/dump.c`,
```c
const char *interactive_option_js = 
#include "utils/mermaid.js"
;
```
In `utils/mermaid.js`,
```js
R""(
const GRAPH_TITLE = new String('graph TB\n');
const FULL_DEPTH = 64;
    ... (put all the remaining contents html related) ...
)""
```

File: cmds/dump.c (hunk 1 above)
Comment (honggyukim) [reply cmds/dump.c:10]: I think it doesn't have to be a header. Please see my comments below.

**Review by honggyukim — COMMENTED** [review]
Thanks for doing this work! I have some requests as follows.

**Comment by honggyukim** [comment]
By the way, I didn't know R string with `R""(...)""` is allowed in C. Is this a standard C? If so, from which version?

Otherwise, this could make build errors in various other environments.

**Comment by wendyisdream** [comment]
Thank you for the review!
I applied the code as you suggested.
For raw string literals,  I can find it was added to C++11 standards. 
But on C standard side, it seems to be just a kind of extension.
I can see gcc support it but clang for C doesn't.
So I changed this option string to C style including escape character for it. (I wanted to avoid this format but I couldn't find any other way in C standard yet)

File: cmds/dump.c [hunk 5]
```diff
@@ -1308,45 +1308,63 @@ static void dump_mermaid_header(struct uftrace_dump_ops *ops, struct uftrace_dat
...
-		pr_out("</div>\n");
-	}
+	pr_out("<h2>Function Call Graph for <span style=\"color:blue\" id=\"baseName\">%s</span>",
+	       mermaid_graph.root.name);
+	pr_out("</h2>\n");
+
+	interactive_option_html =
+#include "utils/mermaid.html"
+		;
+	pr_out("%s", interactive_option_html);
```
Comment (namhyung) [inline cmds/dump.c:1346]: Maybe you can make it a normal html file then copy the file directly (using `fwrite(..., outfp)`?)

File: cmds/dump.c [hunk 6]
```diff
@@ -1308,45 +1308,63 @@ static void dump_mermaid_header(struct uftrace_dump_ops *ops, struct uftrace_dat
...
+	pr_out("	maxTextSize: 99999999,\n");
+	pr_out("	flowchart: { useMaxWidth: false, curve : \'basis\', htmlLabels: \'false\'},\n");
+	pr_out("	securityLevel:\'loose\'\n");
+	pr_out("};\n");
+	pr_out("mermaid.initialize(config);\n");
+
+	interactive_option_js =
+#include "utils/mermaid.js"
+		;
+	pr_out("%s", interactive_option_js);
```
Comment (namhyung) [inline cmds/dump.c:1365]: Ditto.

File: cmds/dump.c (hunk 5 above)
Comment (honggyukim) [reply cmds/dump.c:1346]: It makes uftrace binary depend on having `mermaid.{html,js}` when running so we chose to do this way.

Comment (wendyisdream) [reply cmds/dump.c:1346]: At first, I also considered using js/html file itself.
But in that case, it is additionally required for uftrace binary to know the path of js/html file because uftrace can be installed everywhere. 
So this PR includes js/html file to uftrace as Hongkki mentioned.

Comment (namhyung) [reply cmds/dump.c:1346]: I see.   Then how about keeping the file as a normal file and convert it to a C file at build time so that we can include it like here?  That'd be better than manually adding escape sequences.  Something like https://stackoverflow.com/a/5500077.

**Comment by wendyisdream** [comment]
Thank you for this suggestion!

I updated the code to keep this js/html and convert it in build time and use the converted file itself.
I also considered converting the js/html file in place.
But I think using the converted file is simple for managing because it doesn't have any effect on the original in case of exception cases like build failure.

File: Makefile [hunk 7]
```diff
@@ -396,4 +400,9 @@ ctags:
 	@find . -name "*\.[chS]" -o -path ./tests -prune -o -path ./check-deps -prune \
 		| xargs ctags --regex-asm='/^(GLOBAL|ENTRY|END)\(([^)]*)\).*/\2/'
 
+c-str-conversion:
+	@for i in ${C_STR_TARGET}; do \
+		sed -e 's/\\/\\\\/g;s/\"/\\"/g;s/$$/\\n\"/;s/^/\"/' $${i} > $${i}.$(C_STR_EXTENSION);\
```
Comment (namhyung) [inline Makefile:405]: Does `s/^/\"/` part really necessary?

Comment (wendyisdream) [reply Makefile:405]: (s/^/\"/) is for adding __"__ at the beginning of the line as the pair of __"__  of (s/$/\\n\"/) at the end of the line like *"          <option value=\"1\">1</option>\n"* for string literal concatenation.


Another way add the backslash at the end of the line and add " for the first and last only.
It will work in the sed command
's/\\/\\\\/g;s/\"/\\"/g;s/$/\\n\\/;1s/^/\"/;$ a \"'.

The current commit applied the string literal concatenation.

Comment (namhyung) [reply Makefile:405]: Ah, I see.  Thanks for the explanation.  Looks good now :)

File: utils/mermaid.html [hunk 8]
```diff
@@ -0,0 +1,57 @@
+<style>
+#options {
+  border-collapse: collapse;
+  width: 100%;
+}
+#options td, #options th {
+  border: 1px solid #ddd;
```
Comment (namhyung) [inline utils/mermaid.html:7]: Is it same as #0d0d0d or #000ddd?

File: utils/mermaid.html [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    <tr>
+      <td>
+          <button class="button" onclick="drawBaseGraph(0)">Full-Graph</button>
+          <button class="button" onclick="drawSelectGraph()">Sub-Graph Selection</button>
+          <span id="description" style="color:blue" > </span>
+      </td>
+      <td>
+        Depth:
+        <select id="depth-select">
+          <option value="64" selected>Full</option>
```
Comment (namhyung) [inline utils/mermaid.html:44]: 64 ?

File: utils/mermaid.js [hunk 10]
```diff
@@ -0,0 +1,192 @@
+const GRAPH_TITLE = new String("graph TB\n");
+const FULL_DEPTH = 64;
```
Comment (namhyung) [inline utils/mermaid.js:2]: Where does this come from?

File: cmds/dump.c [hunk 11]
```diff
@@ -1308,45 +1308,65 @@ static void dump_mermaid_header(struct uftrace_dump_ops *ops, struct uftrace_dat
...
+	pr_out("</h2>\n");
+
+	interactive_option_html =
+#include "utils/mermaid.html.cstr" /* This file is a converted of mermaid.html to one string literal in build-time */
+		;
+
+	pr_out("%s", interactive_option_html);
+
+	pr_out("<div class=\"mermaid\" id=\"meraid_graph\">\n");
+	pr_out("flowchart TB\n");
```
Comment (namhyung) [inline cmds/dump.c:1350]: What is TB?

File: utils/mermaid.js [hunk 12]
```diff
@@ -0,0 +1,192 @@
...
+		callFlow[i] = callFlow[i].replace(COMMENT, "");
+		if (Number(startDepth[i]) >= depth)
+			callFlow[i] = (COMMENT + callFlow[i]);
+		if (Number(startDepth[i]) == depth) {
+			id = startDepth[i] + "_" + startId[i];
+			if (nodeList.includes(id) == false) {
+				nodeList.push(id);
+				showHideDesc = showHideDesc.concat(
+					"  style " + id +
+					" fill:#f9f,stroke:#333,stroke-width:px;\n");
```
Comment (namhyung) [inline utils/mermaid.js:91]: Can we move this style change to a separate place?

File: utils/mermaid.js [hunk 13]
```diff
@@ -0,0 +1,192 @@
...
+	let selectionDesc = new String();
+	let nodeList = new Array();
+	if (callFlow.length == 0)
+		drawBaseGraph(0);
+	for (let i = 1; i < callFlow.length; i++) {
+		let id = startDepth[i] + "_" + startId[i];
+		if (nodeList.includes(id) == false) {
+			nodeList.push(id);
+			selectionDesc = selectionDesc.concat(
+				"  style " + id + " fill:#CAB8FF,stroke:#333,stroke-width:px;\n");
```
Comment (namhyung) [inline utils/mermaid.js:110]: Ditto.

File: utils/mermaid.js [hunk 14]
```diff
@@ -0,0 +1,192 @@
...
+	mermaid.initialize(config);
+	initializeBaseGraph();
+	let start = new Boolean(false);
+	maxDepth = 0;
+	var result = mermaidTxt.split(/\r?\n/);
+	for (let i = 0; i < result.length; i++) {
+		if (result[i].length == 0)
+			continue;
+		let oneCallFlow = result[i].match(
+			/\s+(?<start_depth>\d+)_(?<start_id>\d+)\[\"(?<start_name>\S+)\"\]\s+--\S+\|(?<call_num>\d+)\|\s+(?<end_depth>\d+)_(?<end_id>\d+)\[\"(?<end_name>\S+)\"];/);
```
Comment (namhyung) [inline utils/mermaid.js:129]: It's a bit hard to follow.  Could you please add a comment with examples for each case?

File: utils/mermaid.js [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+	baseGraph = "".concat(GRAPH_TITLE, callFlow.join("\n"));
+	updateDepthList();
+	renderGraph(baseGraph);
+};
+
+var updateDepthList =
+	function() {
+	var select = document.getElementById("depth-select");
+	select.options.length = 0;
+	select.options[0] = new Option("Full", 64);
```
Comment (namhyung) [inline utils/mermaid.js:161]: FULL_DEPTH ?

**Review by namhyung — COMMENTED** [review]
Sorry for the late review.

File: utils/mermaid.html (hunk 8 above)
Comment (wendyisdream) [reply utils/mermaid.html:7]: It means #dddddd.
Please refer to the below link's [3 Digit HEX Value] part.
https://www.w3schools.com/css/css_colors_hex.asp

File: utils/mermaid.html (hunk 9 above)
Comment (wendyisdream) [reply utils/mermaid.html:44]: 64 is the value I set to the initial call depth limit.
I just think there is no normal case to be over 64 call depth.

File: utils/mermaid.js (hunk 10 above)
Comment (wendyisdream) [reply utils/mermaid.js]: I updated to use the value in Javascript as HTML's initial value for "full".

File: cmds/dump.c (hunk 11 above)
Comment (wendyisdream) [reply cmds/dump.c:1353]: TB means the flowchart is oriented from Top to Bottom. '
(The other option is LR(Left to Right)])
Please refer to the below link.
https://mermaid-js.github.io/mermaid/#/flowchart

File: utils/mermaid.js (hunk 12 above)
Comment (wendyisdream) [reply utils/mermaid.js]: I updated it to use the separate funtion.

File: utils/mermaid.js (hunk 13 above)
Comment (wendyisdream) [reply utils/mermaid.js]: I updated it to use the separate function.

File: utils/mermaid.js (hunk 14 above)
Comment (wendyisdream) [reply utils/mermaid.js:135]: I added the comment of an example.

File: utils/mermaid.js (hunk 15 above)
Comment (wendyisdream) [reply utils/mermaid.js]: Yes. I updated it to use the defined constant.
Thank you for the review.

File: utils/mermaid.html (hunk 8 above)
Comment (namhyung) [reply utils/mermaid.html:7]: Ok, thanks!

File: utils/mermaid.html (hunk 9 above)
Comment (namhyung) [reply utils/mermaid.html:44]: You don't know how much it would use.  The default max call depth in uftrace is 1024 as C++ programs tend to have many small functions.  But practically we won't allow that many in the display..

File: utils/mermaid.js [hunk 16]
```diff
@@ -0,0 +1,202 @@
...
+	for (let i = 0; i < callFlow.length; i++) {
+		callFlow[i] = callFlow[i].replace(COMMENT, "");
+		if (Number(startDepth[i]) >= depth)
+			callFlow[i] = (COMMENT + callFlow[i]);
+		if (Number(startDepth[i]) == depth) {
+			id = startDepth[i] + "_" + startId[i];
+			if (nodeList.includes(id) == false) {
+				nodeList.push(id);
+				showHideDesc =
+					showHideDesc.concat(getNodeStyle(id, "#f9f", "#333"), "\n");
```
Comment (namhyung) [inline utils/mermaid.js:95]: It'd be nice if we have a symbolic name for this constant so that one can customize it easily.

File: utils/mermaid.js [hunk 17]
```diff
@@ -0,0 +1,202 @@
...
+	let selectionDesc = new String();
+	let nodeList = new Array();
+	if (callFlow.length == 0)
+		drawBaseGraph(0);
+	for (let i = 1; i < callFlow.length; i++) {
+		let id = startDepth[i] + "_" + startId[i];
+		if (nodeList.includes(id) == false) {
+			nodeList.push(id);
+			selectionDesc =
+				selectionDesc.concat(getNodeStyle(id, "#cab8ff", "#333"), "\n");
```
Comment (namhyung) [inline utils/mermaid.js:115]: Ditto.  Please use symbolic names for the hard coded values.

File: utils/mermaid.js (hunk 14 above)
Comment (namhyung) [reply utils/mermaid.js:134]: Thanks!

**Review by namhyung — COMMENTED** [review]
Thanks for the update!

File: utils/mermaid.js (hunk 16 above)
Comment (wendyisdream) [reply utils/mermaid.js]: Thank you for the information!
I updated it to 1024

And I changed this color value to use a constant definition. It looks cleaner. Thank you.

**Review by namhyung — APPROVED** [review]
LGTM

File: Makefile (hunk 7 above)
Comment (honggyukim) [reply Makefile:405]: I would like to ask a small cosmetic change request as follows.
```suggestion
		sed -e 's#\\#\\\\#g;s#\"#\\"#g;s#$$#\\n\"#;s#^#\"#' $${i} > $${i}.$(C_STR_EXTENSION); \
```
It's to make less confusing by using `#` as a separator instead of `/` because using `/` and `\` together looks confusing.

**Review by honggyukim — COMMENTED** [review]
Sorry, I couldn't take a look for a while but left a little nitpick change request.

**Comment by wendyisdream** [comment]
> Sorry, I couldn't take a look for a while but left a little nitpick change request.

I updated the code as you suggested. Thank you!

**Review by honggyukim — APPROVED** [review]
Thanks for the update. LGTM.
